pkg/tuned/controller.go

```diff
@@ -1250,6 +1324,77 @@ func (c *Controller) updateTunedProfileStatus(ctx context.Context, change Change
 	return nil
 }
 
+// storeDeferredUpdate sets the node state (on storage, like disk) to signal
+// there is a deferred update pending.
+func (c *Controller) storeDeferredUpdate(deferredFP string) (derr error) {
+	// "overwriting" is fine, because we only want an empty data file.
+	// all the races are benign, so we go for the simplest approach
+	fp, err := os.Create(tunedDeferredUpdateEphemeralFilePath)
+	if err != nil {
+		return err
+	}
+	_ = fp.Close() // unlikely to fail, we don't write anything
+
+	// overwriting files is racy, and output can be mixed in.
+	// the safest approach is to create a temporary file, write
+	// the full content to it and then rename it, because rename(2)
+	// is atomic, this is guaranteed safe and race-free.
+	dst, err := os.CreateTemp(ocpTunedHome, "ocpdeferred")
+	if err != nil {
+		return err
+	}
+	tmpName := dst.Name()
+	defer func() {
+		if dst == nil {
+			return
+		}
+		derr = dst.Close()
+		os.Remove(dst.Name()) // avoid littering with tmp files
+	}()
+	if _, err := dst.WriteString(deferredFP); err != nil {
+		return err
+	}
+	if err := dst.Close(); err != nil {
+		return err
+	}
+	dst = nil // avoid double close()s, the second will fail
+	return os.Rename(tmpName, tunedDeferredUpdatePersistentFilePath)
+}
+
+// recoverAndClearDeferredUpdate detects the presence and removes the persistent
+// deferred updates file.
+// Returns:
+//   - The defered profile fingerprint.
+//   - A boolean indicating the absence of the ephemeral deferred update file.
+//     If the file is absent, a node restart is assumed and true is returned.
+//   - Error if any.
+func (c *Controller) recoverAndClearDeferredUpdate() (string, bool, error) {
+	isReboot := false
+
+	deferredFP, err := os.ReadFile(tunedDeferredUpdatePersistentFilePath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			klog.Infof("recover: no pending deferred change")
+			return "", isReboot, nil
+		}
+		klog.Infof("recover: failed to restore pending deferred change: %v", err)
+		return "", isReboot, err
+	}
+	pendingFP := strings.TrimSpace(string(deferredFP))
+	err = os.Remove(tunedDeferredUpdatePersistentFilePath)
+	klog.Infof("recover: found pending deferred update: %q", pendingFP)
+
+	if _, errEph := os.Stat(tunedDeferredUpdateEphemeralFilePath); errEph != nil {
```

Comment:
This block can be aligned to left:
```
if _, errEph := os.Stat(tunedDeferredUpdateEphemeralFilePath); errEph != nil {
  if os.IsNotExist(errEph) {
   return pendingFP, true, err
 }
  klog.Infof("recover: failed to detect node restart, assuming not: %v", err)
  return "", false, errEph
}
return pendingFP, isReboot, err

Reply:
good catch, I'll address if I have to resubmit (which is likely anyway :\)

Reply:
done in https://github.com/openshift/cluster-node-tuning-operator/pull/1118